waspc/cli/Wasp/Cli/Command/Watch.hs

```diff
@@ -29,25 +31,37 @@ import qualified Wasp.Lib
 
 -- | Forever listens for any file changes in waspProjectDir, and if there is a change,
 --   compiles Wasp source files in waspProjectDir and regenerates files in outDir.
+--   It will defer recompilation until no change was detected in the last second.
 watch :: Path' Abs (Dir Common.WaspProjectDir) -> Path' Abs (Dir Wasp.Lib.ProjectRootDir) -> IO ()
 watch waspProjectDir outDir = FSN.withManager $ \mgr -> do
   currentTime <- getCurrentTime
   chan <- newChan
   _ <- FSN.watchDirChan mgr (SP.fromAbsDir waspProjectDir) eventFilter chan
   _ <- FSN.watchTreeChan mgr (SP.fromAbsDir $ waspProjectDir </> Common.extCodeDirInWaspProjectDir) eventFilter chan
-  listenForEvents chan currentTime
+  listenForEvents chan currentTime False
   where
-    listenForEvents :: Chan FSN.Event -> UTCTime -> IO ()
-    listenForEvents chan lastCompileTime = do
-      event <- readChan chan
-      let eventTime = FSN.eventTime event
-      if eventTime < lastCompileTime
-        then -- If event happened before last compilation started, skip it.
-          listenForEvents chan lastCompileTime
-        else do
+    listenForEvents :: Chan FSN.Event -> UTCTime -> Bool -> IO ()
```

Comment:
Looking good! Look at the algorithm, it seems to me like it should work ok.
I did find it a bit hard to follow though, in the sense that control flow is a bit complex -> depending on value of `recompileAfterPreemptibleDelay` and on the value returned by the `race`, there are different things that can happen. Due to `recompileAfterPreemptibleDelay` the function effectively operates in two different modes. I wonder if we could somehow split the code into multiple functions to make this a bit more explicit / simpler?

I will give it a quick try here, in pseudocode:
```
- listenForEvents (fn)
  - wait for event
  - if an old event
    - continue from start (listenForEvents)
    else
    - race (wait for another event) (1s delay)
    - if another event && is new enough
      - go back to `race` above
      else
      - recompile
      - go to start (listenForEvents)
```   
     
It is the same thing really, but I think it is a bit clearer regarding the control flow?
As for repeating the race again and again, I would just create another function called `waitUntilNoEvents` or smth like that (there is probably a better name) that takes duration and basically blocks until no events happened for that duration -> it would be this part of logic that calls `race` again and again.
And then that would be called from the `listenForEvents`.

So that would then be

```
- listenForEvents (fn)
  - wait for event
  - if an old event
    - continue from start (listenForEvents)
    else
    - waitUntilNoEvents 1s
    - recompile
    - go to start (listenForEvents)

- waitUntilNoEvents (fn)
  - race (wait for another event) (1s delay)
    - if another event && is new enough
      - go to start (waitUntilNoEvents)
      else
      - done
```   

Also I see that I used 2 times the concept of "event is old/young" in pseudocode -> maybe that could also be made into a separate function? Not sure though.

Reply:
Thanks for the thoughts on trying to simplify the logic. I will review this sketch and test it out in code and see how it looks/feels! Appreciate it 👍🏻

Reply:
Np!
Btw while reviewing the code I remembered one thing from the Clean Code book -> it says that functions taking Bool as an argument are always suspects for doing too much :D. And it refers to the case when Bool is used to control the flow of the function. Recommended "fix" is to try and split function into multiple functions. So that immediately caught my eye, and I think is a useful rule (if not followed blindly of course).

Reply:
Yeah, it is a good call to not allow the function to get too split brained for sure!

Reply:
Refactor here: https://github.com/wasp-lang/wasp/pull/403/commits/ccca4fe67717314a03e83407f8004a9344f57d4f Agree much easier to follow separated out, thnx for the suggestion, will keep my eyes peeled for those opportunities more moving forward :D